docs/electrical/embedded/styleguide.md

```diff
@@ -1,3 +1,190 @@
+2. Pointer variables should be declared on a line separate from non-pointer variables of the same type, to avoid confusion between the two.
+
+#### Alignment Rules
+
+1. TODO: Consider whether to always left align the # of a preprocessor directive, regardless of the indentation of the line
+2. Align relevant blocks of code together (Don't be stupid) 
+    TODO: Make this nicer?
+
+#### Tab Rules
+
+1. The unicode tab character should not appear in files. Instead, the specified tab size is 4 spaces.
+
+### Module Rules
+
+#### Naming Conventions
+
+1. File names should consist only of lower case letters, numbers, and underscores.
+2. File names are recommended to be no longer than 16 characters. Ideally file names are specific and short, like the file itself.
+3. All module names should not overlap with those of the standard library.
+4. All files that contain a ```main()``` function should include "main" within the filename.
+
+#### Header Files
+
+1. Include preprocessor guards against multiple inclusion within all header files (#ifndef NAME_H)
+2. No variables should be declared within a header file, nor should any storage be allocated
+3. Header files should concern themselves strictly with the information that is needed to be known by other modules. (Everything else should be declared within the source file itself)
+
+#### Source Files
+
+1. Source files should be specific, and contain only the information relevant to controlling one 'domain' (UART driver, Servo control, etc).
+2. Source files should always ```#include``` the corresponding header file.
+3. No absolute paths should be used in ```#include``` statements
+4. Do not include extraneous ```#include``` statements.
+5. Do not directly ```#include``` another source file; always ```#include``` the relevant header file.
+6. Any variables shared internally across the entire module but that aren't declared according to the rules of the header file should be declared at the top of the module's source file.
+
+### Data Type Rules
+
+#### Type Definitions
+
+1. When defining new data types, they should always be appended with the ```_t``` designation
+2. Data types should be defined only through ```typedef``` declarations
+3. Any data types that are ```public``` should have their module name prepended to them in the pattern ```module_type```...
+
+#### Use of Types
+
+1. Any time a specific number of bits needs to be used, one of the fixed-width integer types defined in ```stdint.h``` should be used, rather than one of the named types (short, long, long long).
+2. ```char``` types should be used **only** when related to strings; See rule 1 in this subsection.
+3. Bit fields--such as those used for register mapping--should be defined only within unsigned integer types.
+4. Bitwise operations should only be done on unsigned integer types.
+5. Signed and unsigned integers should preferably never be mixed within operations
+6. Avoid floating point operations wherever possible.
```

Comment:
This is a good general rule, but I might say whenever *practical* or *necessary*.

There are lots of cases where it is "possible" to avoid floating points, but it is impractical (little benefit, counterproductive, produces worse code, violates common mathematical assumptions). Likewise, it may be unnecessary (if you have an FPU, the code using floats runs infrequently, or you're writing code that is disabled in production eg debug code).